Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -273,18 +273,25 @@ package final actor SemanticIndexManager {
   /// build system that don't currently have a unit with a timestamp that matches the mtime of the file.
   ///
   /// If `filesToIndex` is `nil`, all files in the build system with out-of-date units are indexed.
+  ///
+  /// If `ensureAllUnitsRegisteredInIndex` is `true`, ensure that all units are registered in the index before
+  /// triggering the indexing. This is a costly operation since it iterates through all the unit files on the file
+  /// system but if existing unit files are not known to the index, we might re-index those files even if they are
+  /// up-to-date. Generally this should be set to `true` during the initial indexing (in which case we might be need to
+  /// build the indexstore-db) and `false` for all subsequent indexing.
   package func scheduleBuildGraphGenerationAndBackgroundIndexAllFiles(
     filesToIndex: [DocumentURI]?,
+    ensureAllUnitsRegisteredInIndex: Bool,
     indexFilesWithUpToDateUnit: Bool
   ) async {
     let taskId = UUID()
     let generateBuildGraphTask = Task(priority: .low) {
       await withLoggingSubsystemAndScope(subsystem: indexLoggingSubsystem, scope: "build-graph-generation") {
         await hooks.buildGraphGenerationDidStart?()
         await self.buildSystemManager.waitForUpToDateBuildGraph()
-        // Ensure that we have an up-to-date indexstore-db. Waiting for the indexstore-db to be updated is cheaper than
-        // potentially not knowing about unit files, which causes the corresponding source files to be re-indexed.
-        index.pollForUnitChangesAndWait()
+        if ensureAllUnitsRegisteredInIndex {
+          index.pollForUnitChangesAndWait()
```

Comment:
Should we add some additional documentation to `pollForUnitChangesAndWait` to describe what it actually does (and the fact it's costly) rather than just "for testing"?

Reply:
Yeah, I’ll add some to indexstore-db